taichi/transforms/type_check.cpp

```diff
@@ -287,7 +324,26 @@ class TypeCheck : public IRVisitor {
       };
       stmt->lhs = promote_custom_int_type(stmt, stmt->lhs);
       stmt->rhs = promote_custom_int_type(stmt, stmt->rhs);
-      auto ret_type = promoted_type(stmt->lhs->ret_type, stmt->rhs->ret_type);
+
+      DataType ret_type;
+      if (is_shift_op(stmt->op_type)) {
+        // shift_ops does not follow the same type promotion rule as numerical
```

Comment:
As shift op is special, why do we still cast rhs to the same type as lhs?

Reply:
Personally I'd prefer making explicit operations at CHI IR level or above, otherwise we have to rely on each specific backend to perform this implicit type cast - their type casting rule may not be consistent with each other.

Btw LLVM doesn't even bother performing this casting:
![image](https://user-images.githubusercontent.com/22334008/165724996-91f680e9-f942-4fac-ad5b-9633ca479f61.png)

Reply:
Thanks for trying this :-) Then let's keep this behavior.